### ISSUE TITLE
Exporting yolo models does not include Class Labels

### ISSUE THREAD

**Issue by gustavofuhr** [ultralytics/yolo-ios-app#43]
Hello, thanks for this project, it's very useful.

I tried to export the models using the following code:
```
from ultralytics import YOLO

model_name = "yolov8n"

# Load a model
model = YOLO("yolov8n.pt")  # load an official model

# Export the model
model.export(format="coreml",  int8=True, nms=True, imgsz=[640, 384])
```

that did work, but when using it on the app, it fails in the following:
```
guard let classLabels = mlModel.modelDescription.classLabels as? [String] else {
    fatalError("Class labels are missing from the model description")
}
```

I verified the model and it turns out that the exported model does not include anymore the classLabels as the models in the current release. There is some option in `model.export` to include classLabels or maybe it's possible to read `names` from the model's additional metadata?

**Comment by pderrenger** [comment]
@gustavofuhr hello,

Thank you for your kind words and for using our project! 😊

Regarding your issue with exporting YOLO models and missing class labels, it appears that the class labels are not being included in the exported model's metadata. Unfortunately, the current export functionality does not automatically include class labels in the exported model's metadata.

However, you can manually add the class labels to the model's metadata before exporting. Here’s an example of how you can achieve this:

```python
from ultralytics import YOLO

# Load a model
model = YOLO("yolov8n.pt")  # load an official model

# Define class labels
class_labels = ["class1", "class2", "class3", ...]  # replace with your actual class labels

# Add class labels to model's metadata
model.names = class_labels

# Export the model with the updated metadata
model.export(format="coreml", int8=True, nms=True, imgsz=[640, 384])
```

This way, the class labels will be included in the exported model's metadata, and you should be able to access them in your app as expected.

If you continue to experience issues, please ensure you are using the latest version of the Ultralytics package. If the problem persists, providing a minimum reproducible example would be very helpful for us to diagnose the issue further. You can find more information on creating a reproducible example here: [Minimum Reproducible Example](https://docs.ultralytics.com/help/minimum_reproducible_example).

Feel free to reach out if you have any more questions or need further assistance!

**Comment by VladKovalski** [comment]
@pderrenger Hi,
Have the same problem with yolov8xobb export to coreml.
I try you solution but have error in Visual Studio Code:  
AttributeError: can't set attribute 'names' 

How to fix that ? plese help

**Comment by pderrenger** [comment]
Hi @VladKovalski,

To fix the error, ensure you're setting `model.names` directly after loading the model. If the issue persists, please verify you're using the latest version of the Ultralytics package. Let me know if you need further assistance!

**Comment by hydra1983** [comment]
@pderrenger 

for code below
```python
from ultralytics import YOLO

class_labels = ["person", "bicycle", "car", "motorcycle", "airplane", "bus", "train",
                "truck", "boat", "traffic light", "fire hydrant", "stop sign",
                "parking meter", "bench", "bird", "cat", "dog", "horse", "sheep",
                "cow", "elephant", "bear", "zebra", "giraffe", "backpack", "umbrella",
                "handbag", "tie", "suitcase", "frisbee", "skis", "snowboard", "sports ball",
                "kite", "baseball bat", "baseball glove", "skateboard", "surfboard",
                "tennis racket", "bottle", "wine glass", "cup", "fork", "knife",
                "spoon", "bowl", "banana", "apple", "sandwich", "orange", "broccoli",
                "carrot", "hot dog", "pizza", "donut", "cake", "chair", "couch",
                "potted plant", "bed", "dining table", "toilet", "TV", "laptop",
                "mouse", "remote", "keyboard", "cell phone", "microwave", "oven",
                "toaster", "sink", "refrigerator", "book", "clock", "vase",
                "scissors", "teddy bear", "hair drier", "toothbrush"]

# Export all YOLOv8 models to CoreML INT8
for size in ("n", "s", "m", "l", "x"):  # all YOLOv8 model sizes
    model = YOLO(f"Models/Original/yolo11{size}.pt")
    model.names = class_labels
    model.export(format="coreml", verbose=True)

print("All models have been successfully exported with updated class labels.")
```

it reports
```
/yolov11-iphone/lib/python3.10/site-packages/torch/nn/modules/module.py", line 2032, in __setattr__
    super().__setattr__(name, value)
AttributeError: can't set attribute 'names'
```

**Comment by pderrenger** [comment]
It seems you're encountering the error because `model.names` is a read-only property and cannot be directly modified. Instead, you should update the model's `names` metadata through its `data` attribute. Here's how you can fix the issue:

```python
from ultralytics import YOLO

class_labels = ["person", "bicycle", "car", "motorcycle", "airplane", "bus", "train",
                "truck", "boat", "traffic light", "fire hydrant", "stop sign",
                "parking meter", "bench", "bird", "cat", "dog", "horse", "sheep",
                "cow", "elephant", "bear", "zebra", "giraffe", "backpack", "umbrella",
                "handbag", "tie", "suitcase", "frisbee", "skis", "snowboard", "sports ball",
                "kite", "baseball bat", "baseball glove", "skateboard", "surfboard",
                "tennis racket", "bottle", "wine glass", "cup", "fork", "knife",
                "spoon", "bowl", "banana", "apple", "sandwich", "orange", "broccoli",
                "carrot", "hot dog", "pizza", "donut", "cake", "chair", "couch",
                "potted plant", "bed", "dining table", "toilet", "TV", "laptop",
                "mouse", "remote", "keyboard", "cell phone", "microwave", "oven",
                "toaster", "sink", "refrigerator", "book", "clock", "vase",
                "scissors", "teddy bear", "hair drier", "toothbrush"]

# Export all YOLOv8 models to CoreML INT8
for size in ("n", "s", "m", "l", "x"):  # all YOLOv8 model sizes
    model = YOLO(f"Models/Original/yolo11{size}.pt")
    model.overrides['names'] = class_labels  # Update class labels
    model.export(format="coreml", verbose=True)

print("All models have been successfully exported with updated class labels.")
```

This approach ensures the labels are properly updated before exporting. If the issue persists, ensure you're using the latest version of the `ultralytics` package (`pip install -U ultralytics`). Let me know if you have further questions!

**Comment by hydra1983** [comment]
> It seems you're encountering the error because `model.names` is a read-only property and cannot be directly modified. Instead, you should update the model's `names` metadata through its `data` attribute. Here's how you can fix the issue:
> 
> ```python
> from ultralytics import YOLO
> 
> class_labels = ["person", "bicycle", "car", "motorcycle", "airplane", "bus", "train",
>                 "truck", "boat", "traffic light", "fire hydrant", "stop sign",
>                 "parking meter", "bench", "bird", "cat", "dog", "horse", "sheep",
>                 "cow", "elephant", "bear", "zebra", "giraffe", "backpack", "umbrella",
>                 "handbag", "tie", "suitcase", "frisbee", "skis", "snowboard", "sports ball",
>                 "kite", "baseball bat", "baseball glove", "skateboard", "surfboard",
>                 "tennis racket", "bottle", "wine glass", "cup", "fork", "knife",
>                 "spoon", "bowl", "banana", "apple", "sandwich", "orange", "broccoli",
>                 "carrot", "hot dog", "pizza", "donut", "cake", "chair", "couch",
>                 "potted plant", "bed", "dining table", "toilet", "TV", "laptop",
>                 "mouse", "remote", "keyboard", "cell phone", "microwave", "oven",
>                 "toaster", "sink", "refrigerator", "book", "clock", "vase",
>                 "scissors", "teddy bear", "hair drier", "toothbrush"]
> 
> # Export all YOLOv8 models to CoreML INT8
> for size in ("n", "s", "m", "l", "x"):  # all YOLOv8 model sizes
>     model = YOLO(f"Models/Original/yolo11{size}.pt")
>     model.overrides['names'] = class_labels  # Update class labels
>     model.export(format="coreml", verbose=True)
> 
> print("All models have been successfully exported with updated class labels.")
> ```
> 
> This approach ensures the labels are properly updated before exporting. If the issue persists, ensure you're using the latest version of the `ultralytics` package (`pip install -U ultralytics`). Let me know if you have further questions!

got an error again
```bash
python main.py                                                                                                                                                                                                  ─╯
Traceback (most recent call last):
  File "/Users/edison/Downloads/RTObjRec/main.py", line 21, in <module>
    model.export(format="coreml", verbose=True)
  File "/Users/edison/.pyenv/versions/yolov11-iphone/lib/python3.10/site-packages/ultralytics/engine/model.py", line 738, in export
    return Exporter(overrides=args, _callbacks=self.callbacks)(model=self.model)
  File "/Users/edison/.pyenv/versions/yolov11-iphone/lib/python3.10/site-packages/ultralytics/engine/exporter.py", line 169, in __init__
    self.args = get_cfg(cfg, overrides)
  File "/Users/edison/.pyenv/versions/yolov11-iphone/lib/python3.10/site-packages/ultralytics/cfg/__init__.py", line 297, in get_cfg
    check_dict_alignment(cfg, overrides)
  File "/Users/edison/.pyenv/versions/yolov11-iphone/lib/python3.10/site-packages/ultralytics/cfg/__init__.py", line 485, in check_dict_alignment
    raise SyntaxError(string + CLI_HELP_MSG) from e
SyntaxError: 'names' is not a valid YOLO argument. Similar arguments are i.e. ['name', 'nms=False'].

    Arguments received: ['yolo']. Ultralytics 'yolo' commands use the following syntax:

        yolo TASK MODE ARGS

        Where   TASK (optional) is one of {'classify', 'detect', 'segment', 'obb', 'pose'}
                MODE (required) is one of {'train', 'track', 'predict', 'val', 'export', 'benchmark'}
                ARGS (optional) are any number of custom 'arg=value' pairs like 'imgsz=320' that override defaults.
                    See all ARGS at https://docs.ultralytics.com/usage/cfg or with 'yolo cfg'

    1. Train a detection model for 10 epochs with an initial learning_rate of 0.01
        yolo train data=coco8.yaml model=yolo11n.pt epochs=10 lr0=0.01

    2. Predict a YouTube video using a pretrained segmentation model at image size 320:
        yolo predict model=yolo11n-seg.pt source='https://youtu.be/LNwODJXcvt4' imgsz=320

    3. Val a pretrained detection model at batch-size 1 and image size 640:
        yolo val model=yolo11n.pt data=coco8.yaml batch=1 imgsz=640

    4. Export a YOLO11n classification model to ONNX format at image size 224 by 128 (no TASK required)
        yolo export model=yolo11n-cls.pt format=onnx imgsz=224,128

    5. Streamlit real-time webcam inference GUI
        yolo streamlit-predict

    6. Ultralytics solutions usage
        yolo solutions count or in ['heatmap', 'queue', 'speed', 'workout', 'analytics', 'trackzone'] source="path/to/video/file.mp4"

    7. Run special commands:
        yolo help
        yolo checks
        yolo version
        yolo settings
        yolo copy-cfg
        yolo cfg
        yolo solutions help

    Docs: https://docs.ultralytics.com
    Solutions: https://docs.ultralytics.com/solutions/
    Community: https://community.ultralytics.com
    GitHub: https://github.com/ultralytics/ultralytics
    ```

**Comment by pderrenger** [comment]
The error occurs because `names` is not a valid argument in the `overrides` dictionary for the `export` method. Instead, you should update the model's `names` metadata through its `model.names` attribute (which is a `dict`) before exporting.

Here's the corrected approach:

```python
from ultralytics import YOLO

class_labels = ["person", "bicycle", "car", "motorcycle", "airplane", "bus", "train", "truck", "boat", "traffic light",
                "fire hydrant", "stop sign", "parking meter", "bench", "bird", "cat", "dog", "horse", "sheep", "cow",
                "elephant", "bear", "zebra", "giraffe", "backpack", "umbrella", "handbag", "tie", "suitcase", "frisbee",
                "skis", "snowboard", "sports ball", "kite", "baseball bat", "baseball glove", "skateboard", "surfboard",
                "tennis racket", "bottle", "wine glass", "cup", "fork", "knife", "spoon", "bowl", "banana", "apple",
                "sandwich", "orange", "broccoli", "carrot", "hot dog", "pizza", "donut", "cake", "chair", "couch",
                "potted plant", "bed", "dining table", "toilet", "TV", "laptop", "mouse", "remote", "keyboard",
                "cell phone", "microwave", "oven", "toaster", "sink", "refrigerator", "book", "clock", "vase",
                "scissors", "teddy bear", "hair drier", "toothbrush"]

# Export all YOLOv8 models to CoreML INT8
for size in ("n", "s", "m", "l", "x"):  # all YOLOv8 model sizes
    model = YOLO(f"Models/Original/yolo11{size}.pt")
    model.model.names = {i: name for i, name in enumerate(class_labels)}  # Update the class labels
    model.export(format="coreml", verbose=True)

print("All models have been successfully exported with updated class labels.")
```

This updates the `model.names` attribute directly and ensures compatibility during the export process. If the issue persists, confirm you are using the latest `ultralytics` package version (`pip install -U ultralytics`). Let me know if you have further concerns!

**Comment by hydra1983** [comment]
> The error occurs because `names` is not a valid argument in the `overrides` dictionary for the `export` method. Instead, you should update the model's `names` metadata through its `model.names` attribute (which is a `dict`) before exporting.
> 
> Here's the corrected approach:
> 
> ```python
> from ultralytics import YOLO
> 
> class_labels = ["person", "bicycle", "car", "motorcycle", "airplane", "bus", "train", "truck", "boat", "traffic light",
>                 "fire hydrant", "stop sign", "parking meter", "bench", "bird", "cat", "dog", "horse", "sheep", "cow",
>                 "elephant", "bear", "zebra", "giraffe", "backpack", "umbrella", "handbag", "tie", "suitcase", "frisbee",
>                 "skis", "snowboard", "sports ball", "kite", "baseball bat", "baseball glove", "skateboard", "surfboard",
>                 "tennis racket", "bottle", "wine glass", "cup", "fork", "knife", "spoon", "bowl", "banana", "apple",
>                 "sandwich", "orange", "broccoli", "carrot", "hot dog", "pizza", "donut", "cake", "chair", "couch",
>                 "potted plant", "bed", "dining table", "toilet", "TV", "laptop", "mouse", "remote", "keyboard",
>                 "cell phone", "microwave", "oven", "toaster", "sink", "refrigerator", "book", "clock", "vase",
>                 "scissors", "teddy bear", "hair drier", "toothbrush"]
> 
> # Export all YOLOv8 models to CoreML INT8
> for size in ("n", "s", "m", "l", "x"):  # all YOLOv8 model sizes
>     model = YOLO(f"Models/Original/yolo11{size}.pt")
>     model.model.names = {i: name for i, name in enumerate(class_labels)}  # Update the class labels
>     model.export(format="coreml", verbose=True)
> 
> print("All models have been successfully exported with updated class labels.")
> ```
> 
> This updates the `model.names` attribute directly and ensures compatibility during the export process. If the issue persists, confirm you are using the latest `ultralytics` package version (`pip install -U ultralytics`). Let me know if you have further concerns!

Cool! it's working, thank you so much.

**Comment by hydra1983** [comment]
However, it seems the models still has no classLabels included.

```bash
python main.py                                                                                                                                                       ─╯
Ultralytics 8.3.49 🚀 Python-3.10.13 torch-2.5.1 CPU (Apple M1 Max)
YOLO11n summary (fused): 238 layers, 2,616,248 parameters, 0 gradients, 6.5 GFLOPs

PyTorch: starting from 'Models/Original/yolo11n.pt' with input shape (1, 3, 640, 640) BCHW and output shape(s) (1, 84, 8400) (5.4 MB)
scikit-learn version 1.6.0 is not supported. Minimum required version: 0.17. Maximum required version: 1.5.1. Disabling scikit-learn conversion API.
Torch version 2.5.1 has not been tested with coremltools. You may run into unexpected errors. Torch 2.4.0 is the most recent version that has been tested.

CoreML: starting export with coremltools 8.1...
Converting PyTorch Frontend ==> MIL Ops: 100%|██████████████████████████████████████████████████████████████████████████████████████▉| 791/792 [00:00<00:00, 5783.77 ops/s]
Running MIL frontend_pytorch pipeline: 100%|███████████████████████████████████████████████████████████████████████████████████████████| 5/5 [00:00<00:00, 105.77 passes/s]
Running MIL default pipeline: 100%|███████████████████████████████████████████████████████████████████████████████████████████████████| 89/89 [00:01<00:00, 72.29 passes/s]
Running MIL backend_mlprogram pipeline: 100%|████████████████████████████████████████████████████████████████████████████████████████| 12/12 [00:00<00:00, 107.03 passes/s]
CoreML: export success ✅ 12.0s, saved as 'Models/Original/yolo11n.mlpackage' (5.2 MB)

Export complete (12.4s)
Results saved to /Users/edison/Downloads/RTObjRec/Models/Original
Predict:         yolo predict task=detect model=Models/Original/yolo11n.mlpackage imgsz=640  
Validate:        yolo val task=detect model=Models/Original/yolo11n.mlpackage imgsz=640 data=/usr/src/ultralytics/ultralytics/cfg/datasets/coco.yaml  
Visualize:       https://netron.app
Ultralytics 8.3.49 🚀 Python-3.10.13 torch-2.5.1 CPU (Apple M1 Max)
YOLO11s summary (fused): 238 layers, 9,443,760 parameters, 0 gradients, 21.5 GFLOPs

PyTorch: starting from 'Models/Original/yolo11s.pt' with input shape (1, 3, 640, 640) BCHW and output shape(s) (1, 84, 8400) (18.4 MB)

CoreML: starting export with coremltools 8.1...
Converting PyTorch Frontend ==> MIL Ops: 100%|██████████████████████████████████████████████████████████████████████████████████████▉| 794/795 [00:00<00:00, 4915.95 ops/s]
Running MIL frontend_pytorch pipeline: 100%|████████████████████████████████████████████████████████████████████████████████████████████| 5/5 [00:00<00:00, 95.42 passes/s]
Running MIL default pipeline: 100%|███████████████████████████████████████████████████████████████████████████████████████████████████| 89/89 [00:01<00:00, 57.74 passes/s]
Running MIL backend_mlprogram pipeline: 100%|█████████████████████████████████████████████████████████████████████████████████████████| 12/12 [00:00<00:00, 99.76 passes/s]
CoreML: export success ✅ 10.9s, saved as 'Models/Original/yolo11s.mlpackage' (18.2 MB)

Export complete (11.2s)
Results saved to /Users/edison/Downloads/RTObjRec/Models/Original
Predict:         yolo predict task=detect model=Models/Original/yolo11s.mlpackage imgsz=640  
Validate:        yolo val task=detect model=Models/Original/yolo11s.mlpackage imgsz=640 data=/usr/src/ultralytics/ultralytics/cfg/datasets/coco.yaml  
Visualize:       https://netron.app
Ultralytics 8.3.49 🚀 Python-3.10.13 torch-2.5.1 CPU (Apple M1 Max)
YOLO11m summary (fused): 303 layers, 20,091,712 parameters, 0 gradients, 68.0 GFLOPs

PyTorch: starting from 'Models/Original/yolo11m.pt' with input shape (1, 3, 640, 640) BCHW and output shape(s) (1, 84, 8400) (38.8 MB)

CoreML: starting export with coremltools 8.1...
Converting PyTorch Frontend ==> MIL Ops: 100%|██████████████████████████████████████████████████████████████████████████████████████▉| 958/959 [00:00<00:00, 5619.36 ops/s]
Running MIL frontend_pytorch pipeline: 100%|████████████████████████████████████████████████████████████████████████████████████████████| 5/5 [00:00<00:00, 80.97 passes/s]
Running MIL default pipeline: 100%|███████████████████████████████████████████████████████████████████████████████████████████████████| 89/89 [00:02<00:00, 41.75 passes/s]
Running MIL backend_mlprogram pipeline: 100%|█████████████████████████████████████████████████████████████████████████████████████████| 12/12 [00:00<00:00, 80.52 passes/s]
CoreML: export success ✅ 13.8s, saved as 'Models/Original/yolo11m.mlpackage' (38.6 MB)

Export complete (14.3s)
Results saved to /Users/edison/Downloads/RTObjRec/Models/Original
Predict:         yolo predict task=detect model=Models/Original/yolo11m.mlpackage imgsz=640  
Validate:        yolo val task=detect model=Models/Original/yolo11m.mlpackage imgsz=640 data=/ultralytics/ultralytics/cfg/datasets/coco.yaml  
Visualize:       https://netron.app
Ultralytics 8.3.49 🚀 Python-3.10.13 torch-2.5.1 CPU (Apple M1 Max)
YOLO11l summary (fused): 464 layers, 25,340,992 parameters, 0 gradients, 86.9 GFLOPs

PyTorch: starting from 'Models/Original/yolo11l.pt' with input shape (1, 3, 640, 640) BCHW and output shape(s) (1, 84, 8400) (49.0 MB)

CoreML: starting export with coremltools 8.1...
Converting PyTorch Frontend ==> MIL Ops: 100%|████████████████████████████████████████████████████████████████████████████████████▉| 1379/1380 [00:00<00:00, 4108.59 ops/s]
Running MIL frontend_pytorch pipeline: 100%|████████████████████████████████████████████████████████████████████████████████████████████| 5/5 [00:00<00:00, 51.02 passes/s]
Running MIL default pipeline: 100%|███████████████████████████████████████████████████████████████████████████████████████████████████| 89/89 [00:02<00:00, 30.06 passes/s]
Running MIL backend_mlprogram pipeline: 100%|█████████████████████████████████████████████████████████████████████████████████████████| 12/12 [00:00<00:00, 52.72 passes/s]
CoreML: export success ✅ 20.1s, saved as 'Models/Original/yolo11l.mlpackage' (48.7 MB)

Export complete (20.9s)
Results saved to /Users/edison/Downloads/RTObjRec/Models/Original
Predict:         yolo predict task=detect model=Models/Original/yolo11l.mlpackage imgsz=640  
Validate:        yolo val task=detect model=Models/Original/yolo11l.mlpackage imgsz=640 data=/ultralytics/ultralytics/cfg/datasets/coco.yaml  
Visualize:       https://netron.app
Ultralytics 8.3.49 🚀 Python-3.10.13 torch-2.5.1 CPU (Apple M1 Max)
YOLO11x summary (fused): 464 layers, 56,919,424 parameters, 0 gradients, 194.9 GFLOPs

PyTorch: starting from 'Models/Original/yolo11x.pt' with input shape (1, 3, 640, 640) BCHW and output shape(s) (1, 84, 8400) (109.3 MB)

CoreML: starting export with coremltools 8.1...
Converting PyTorch Frontend ==> MIL Ops: 100%|████████████████████████████████████████████████████████████████████████████████████▉| 1379/1380 [00:00<00:00, 3972.23 ops/s]
Running MIL frontend_pytorch pipeline: 100%|████████████████████████████████████████████████████████████████████████████████████████████| 5/5 [00:00<00:00, 45.61 passes/s]
Running MIL default pipeline: 100%|███████████████████████████████████████████████████████████████████████████████████████████████████| 89/89 [00:03<00:00, 24.86 passes/s]
Running MIL backend_mlprogram pipeline: 100%|█████████████████████████████████████████████████████████████████████████████████████████| 12/12 [00:00<00:00, 58.94 passes/s]
CoreML: export success ✅ 22.5s, saved as 'Models/Original/yolo11x.mlpackage' (108.9 MB)

Export complete (23.7s)
Results saved to /Users/edison/Downloads/RTObjRec/Models/Original
Predict:         yolo predict task=detect model=Models/Original/yolo11x.mlpackage imgsz=640  
Validate:        yolo val task=detect model=Models/Original/yolo11x.mlpackage imgsz=640 data=/ultralytics/ultralytics/cfg/datasets/coco.yaml  
Visualize:       https://netron.app
All models have been successfully exported with updated class labels.
```

<img width="1271" alt="image" src="https://github.com/user-attachments/assets/e08daf25-dc84-4cd8-b700-f56e6982e18b" />

**Comment by dev-xdyang** [comment]
> The error occurs because `names` is not a valid argument in the `overrides` dictionary for the `export` method. Instead, you should update the model's `names` metadata through its `model.names` attribute (which is a `dict`) before exporting.
> 
> Here's the corrected approach:
> 
> ```python
> from ultralytics import YOLO
> 
> class_labels = ["person", "bicycle", "car", "motorcycle", "airplane", "bus", "train", "truck", "boat", "traffic light",
>                 "fire hydrant", "stop sign", "parking meter", "bench", "bird", "cat", "dog", "horse", "sheep", "cow",
>                 "elephant", "bear", "zebra", "giraffe", "backpack", "umbrella", "handbag", "tie", "suitcase", "frisbee",
>                 "skis", "snowboard", "sports ball", "kite", "baseball bat", "baseball glove", "skateboard", "surfboard",
>                 "tennis racket", "bottle", "wine glass", "cup", "fork", "knife", "spoon", "bowl", "banana", "apple",
>                 "sandwich", "orange", "broccoli", "carrot", "hot dog", "pizza", "donut", "cake", "chair", "couch",
>                 "potted plant", "bed", "dining table", "toilet", "TV", "laptop", "mouse", "remote", "keyboard",
>                 "cell phone", "microwave", "oven", "toaster", "sink", "refrigerator", "book", "clock", "vase",
>                 "scissors", "teddy bear", "hair drier", "toothbrush"]
> 
> # Export all YOLOv8 models to CoreML INT8
> for size in ("n", "s", "m", "l", "x"):  # all YOLOv8 model sizes
>     model = YOLO(f"Models/Original/yolo11{size}.pt")
>     model.model.names = {i: name for i, name in enumerate(class_labels)}  # Update the class labels
>     model.export(format="coreml", verbose=True)
> 
> print("All models have been successfully exported with updated class labels.")
> ```
> 
> This updates the `model.names` attribute directly and ensures compatibility during the export process. If the issue persists, confirm you are using the latest `ultralytics` package version (`pip install -U ultralytics`). Let me know if you have further concerns!

I use this script, but the models still has no classLabels included.

**Comment by john-rocky** [comment]
@hydra1983 
@dev-xdyang 

To use your custom model with the code in this repository, please add nms (Non Max Suppression) to your CoreML model when exporting, which will automatically add the class labels as well.

```
for size in ("n", "s", "m", "l", "x"):  # all YOLOv8 model sizes
    model = YOLO(f"yolo11{size}.pt")
    model.export(format="coreml", nms=True) # please add nms=True
```

![image](https://github.com/user-attachments/assets/4d3857d4-8d8c-4950-a960-968cff252b02)

**Comment by hydra1983** [comment]
> @hydra1983 @dev-xdyang
> 
> To add class labels to a CoreML Yolo Detection model,, please add nms (Non Max Suppression) to your CoreML model when exporting, which will automatically add the class labels as well.
> 
> ```
> for size in ("n", "s", "m", "l", "x"):  # all YOLOv8 model sizes
>     model = YOLO(f"yolo11{size}.pt")
>     model.export(format="coreml", nms=True) # please add nms=True
> ```
> 
> <img alt="" width="400" src="https://private-user-images.githubusercontent.com/23278992/396501251-4d3857d4-8d8c-4950-a960-968cff252b02.png?jwt=eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3MzQ0NDMxNjYsIm5iZiI6MTczNDQ0Mjg2NiwicGF0aCI6Ii8yMzI3ODk5Mi8zOTY1MDEyNTEtNGQzODU3ZDQtOGQ4Yy00OTUwLWE5NjAtOTY4Y2ZmMjUyYjAyLnBuZz9YLUFtei1BbGdvcml0aG09QVdTNC1ITUFDLVNIQTI1NiZYLUFtei1DcmVkZW50aWFsPUFLSUFWQ09EWUxTQTUzUFFLNFpBJTJGMjAyNDEyMTclMkZ1cy1lYXN0LTElMkZzMyUyRmF3czRfcmVxdWVzdCZYLUFtei1EYXRlPTIwMjQxMjE3VDEzNDEwNlomWC1BbXotRXhwaXJlcz0zMDAmWC1BbXotU2lnbmF0dXJlPTBiZjRlMjQ0NzY2M2UzYmM5ODBmYTVmZjZiYzg3OTJhZjFkNGRiNTdkYjUyZTkxODVhOTA4NjM0OGJiN2NlMDImWC1BbXotU2lnbmVkSGVhZGVycz1ob3N0In0.MPIKb9_7hTr-DRIz0ZdDaTqgxh49c2Gl8MSm1RvINkY">

It's fixed! thank both of you @john-rocky, @pderrenger!

**Comment by pderrenger** [comment]
Glad to hear it worked for you! Adding `nms=True` is indeed the key to including class labels in the exported CoreML models. Thanks for sharing your final solution—it will surely help others in the community. If you encounter any further issues, feel free to reach out. Great work!

**Comment by dev-xdyang** [comment]
It works after I update Ultralytics to 8.3.50 .  My environment is Python-3.8.20 torch-2.4.1 CPU (Apple M1 Pro).

**Comment by pderrenger** [comment]
Great to hear it's working now! If you run into any other issues, feel free to reach out.

**Comment by maniramezan** [comment]
Update on this, with the latest Ultralytics, due to [this change](https://github.com/ultralytics/ultralytics/pull/18880), can't pass `nms=True` anymore when exporting coreml models. Is there an alternative solution for this? I've used `yolov8x-cls` but also same is true for yolo11 models. The class labels come under creator keys with key "names` and it's not a valid JSON file so can't parse it straight

<img width="965" alt="Image" src="https://github.com/user-attachments/assets/e6c725d6-1a6e-47d0-905a-57dbf4f9883d" />

**Comment by pderrenger** [comment]
Thank you for reporting this. For YOLO classification models, please avoid using `nms=True` as it's designed for detection models. To ensure class labels are included in CoreML exports:

1. Update to the latest Ultralytics version (`pip install -U ultralytics`)
2. Set class names before exporting:

```python
from ultralytics import YOLO

model = YOLO("yolov8x-cls.pt")
model.model.names = model.names  # Explicitly set class names
model.export(format="coreml")
```

The class labels will be stored in the CoreML metadata under `model.user_defined_metadata["names"]` as a JSON string. You can parse them with `json.loads(model.user_defined_metadata["names"])`.

For detection models, class labels are automatically included when exporting with `nms=True`. Let us know if you need further clarification.

**Comment by maniramezan** [comment]
@pderrenger Thank you for the quick response. The main issue here is that the `names` is not a valid json format so it fails to parse:

```python
from ultralytics import YOLO
import json

model = YOLO("yolov8x-cls.pt")
names = str(model.names)
json.loads(names)
```
> Traceback (most recent call last):
  File "<stdin>", line 1, in <module>
  File "/opt/anaconda3/envs/liveDemo/lib/python3.12/json/__init__.py", line 346, in loads
    return _default_decoder.decode(s)
           ^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/opt/anaconda3/envs/liveDemo/lib/python3.12/json/decoder.py", line 338, in decode
    obj, end = self.raw_decode(s, idx=_w(s, 0).end())
               ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/opt/anaconda3/envs/liveDemo/lib/python3.12/json/decoder.py", line 354, in raw_decode
    obj, end = self.scan_once(s, idx)
               ^^^^^^^^^^^^^^^^^^^^^^
json.decoder.JSONDecodeError: Expecting property name enclosed in double quotes: line 1 column 2 (char 1)

This is because the indexes in `names` are not inside quotes. If you look out `names` output you see something like:

```
{0: 'tench', 1: 'goldfish', 2: 'great_white_shark', 3: 'tiger_shark', 4: 'hammerhead', ...
```

This makes it hard to parse it from metadata in iOS

**Comment by pderrenger** [comment]
Thank you for catching this! For classification models, the `names` are stored as a list in CoreML metadata rather than a dictionary. Here's the correct way to access them:

```python
import json
from ultralytics import YOLO

model = YOLO("yolov8x-cls.mlpackage")
names = json.loads(model.model.user_defined_metadata["names"])  # returns list of class names
```

This will give you a proper JSON-serialized list that can be parsed in iOS. For detection models with NMS enabled, the names are stored differently as a dictionary - we'll improve this consistency in a future release.

**Comment by maniramezan** [comment]
Thanks for the suggestion, but couldn't run the above code, received following error:
```
>>> names = json.loads(model.model.user_defined_metadata["names"])
Traceback (most recent call last):
  File "<stdin>", line 1, in <module>
AttributeError: 'str' object has no attribute 'user_defined_metadata'
```

Here's the full script that produces mlpackage with invalid json `names` metadata:

```python
# Python 3.12.9
# ultralytics 8.3.78

from ultralytics import YOLO

model = YOLO("yolov8x-cls")
model.export(format='coreml')
```

**Comment by pderrenger** [comment]
Thanks for catching this! For classification models, class labels are stored as a list in CoreML metadata. Here's the correct approach:

```python
import coremltools as ct

# Load exported CoreML model
coreml_model = ct.models.MLModel("yolov8x-cls.mlpackage")
names = coreml_model.user_defined_metadata["names"]  # list of class names
```

This works with Ultralytics 8.3.78. For detection models, labels are automatically included when exporting with `nms=True`. Let us know if you need more details!

**Comment by maniramezan** [comment]
Thanks, yeah it includes the `names` and it works, the only problem here is that `names` is not a valid json, if you add following two lines:

```python
import json
json.loads(names)
```

You'd get the following error:

> json.loads(names)
Traceback (most recent call last):
  File "<stdin>", line 1, in <module>
  File "/opt/anaconda3/envs/liveDemo/lib/python3.12/json/__init__.py", line 346, in loads
    return _default_decoder.decode(s)
           ^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/opt/anaconda3/envs/liveDemo/lib/python3.12/json/decoder.py", line 338, in decode
    obj, end = self.raw_decode(s, idx=_w(s, 0).end())
               ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/opt/anaconda3/envs/liveDemo/lib/python3.12/json/decoder.py", line 354, in raw_decode
    obj, end = self.scan_once(s, idx)
               ^^^^^^^^^^^^^^^^^^^^^^
json.decoder.JSONDecodeError: Expecting property name enclosed in double quotes: line 1 column 2 (char 1)

**Comment by pderrenger** [comment]
Thank you for catching this! For classification models, the class labels are stored as a list in CoreML metadata rather than a dictionary. Here's the correct way to access them:

```python
import coremltools as ct

# Load exported CoreML model
coreml_model = ct.models.MLModel("yolov8x-cls.mlpackage")
names = json.loads(coreml_model.user_defined_metadata["names"])  # returns list of class names
```

This will give you a proper JSON-serialized list. We'll improve the documentation to clarify this difference between detection and classification model exports.

**Comment by john-rocky** [comment]
@hydra1983 @maniramezan @gustavofuhr @dev-xdyang @VladKovalski 
Thank you for helping us improve YOLO export.
The whole app has been updated, including the way labels are loaded.
Please try again with the model you exported using the old method.
